File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter13/ManipulationManagerDemo/ManipulationManagerDemo/ManipulationManager.cpp
#include "pch.h"
#include "ManipulationManager.h"

using namespace ManipulationManagerDemo;

using namespace Platform;
using namespace Windows::Foundation;
using namespace Windows::UI::Input;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Input;
using namespace Windows::UI::Xaml::Media;

ManipulationManager::ManipulationManager()
{
    xformGroup = ref new TransformGroup();
    matrixXform = ref new MatrixTransform();
    xformGroup->Children->Append(matrixXform);
    compositeXform = ref new CompositeTransform();
    xformGroup->Children->Append(compositeXform);
    matrix = MatrixHelper::Identity;
}

void ManipulationManager::AccumulateDelta(Point position, ManipulationDelta delta)
{
    matrixXform->Matrix = xformGroup->Value;
    Point center = matrixXform->TransformPoint(position);
    compositeXform->CenterX = center.X;
    compositeXform->CenterY = center.Y;
    compositeXform->TranslateX = delta.Translation.X;
    compositeXform->TranslateY = delta.Translation.Y;
    compositeXform->ScaleX = delta.Scale;
    compositeXform->ScaleY = delta.Scale;
    compositeXform->Rotation = delta.Rotation;
    matrix = xformGroup->Value;
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter13/ManipulationManagerDemo/ManipulationManagerDemo/ManipulationManager.cs
using Windows.Foundation;
using Windows.UI.Input;
using Windows.UI.Xaml.Media;

namespace ManipulationManagerDemo
{
    public class ManipulationManager
    {
        TransformGroup xformGroup;
        MatrixTransform matrixXform;
        CompositeTransform compositeXform;

        public ManipulationManager()
        {
            xformGroup = new TransformGroup();
            matrixXform = new MatrixTransform();
            xformGroup.Children.Add(matrixXform);
            compositeXform = new CompositeTransform();
            xformGroup.Children.Add(compositeXform);
            this.Matrix = Matrix.Identity;
        }

        public Matrix Matrix { private set; get; }

        public void AccumulateDelta(Point position, ManipulationDelta delta)
        {
            matrixXform.Matrix = xformGroup.Value;
            Point center = matrixXform.TransformPoint(position);
            compositeXform.CenterX = center.X;
            compositeXform.CenterY = center.Y;
            compositeXform.TranslateX = delta.Translation.X;
            compositeXform.TranslateY = delta.Translation.Y;
            compositeXform.ScaleX = delta.Scale;
            compositeXform.ScaleY = delta.Scale;
            compositeXform.Rotation = delta.Rotation;
            this.Matrix = xformGroup.Value;
        }
    }
}
